### ISSUE TITLE
The predicted value is always a straight line

### ISSUE THREAD

**Issue by ZhuzengChu** [MingjunZhong/seq2point-nilm#3]
Hi,@MingjunZhong 
Thank you for the source code.I ran into some problems.
![Figure_1](https://user-images.githubusercontent.com/58038343/88023920-7480fd00-cb64-11ea-9894-343bb6c6c1c9.png)
![Figure_2](https://user-images.githubusercontent.com/58038343/88023934-7a76de00-cb64-11ea-9561-917c76855bb9.png)
There are several devices that have a true value of zero,Is there something wrong with the data normalization?
Thanks.

**Comment by MingjunZhong** [comment]
@ZhuzengChu  Sorry for delay. I cannot justify what is happening. What I suggest is to predict the training data. That means, you apply your trained model to a sequence taken from your training data. This is important because if the performance is poor the model was not well trained, if the performance is good, the model was properly trained. Can you let me know the results?

**Comment by ZhuzengChu** [comment]
@MingjunZhong Thank you for your reply.What you said makes sense. To understand the effect of the test set prediction, must first understand the effect of the training set.Figure 1 is the result of using the training set on seq2point.Figure 2 is the result of using the training set on transfer_learning. The appliance is fridge. 
![Figure_1](https://user-images.githubusercontent.com/58038343/88286567-878dfb80-cd23-11ea-8299-80a49e485a9a.png)
![Figure_2](https://user-images.githubusercontent.com/58038343/88286583-8c52af80-cd23-11ea-88c8-9e29c57869b3.png)
From the prediction result of the training set, the prediction result of the test set is correct. Are MAE and SAE more important in the NILM indicators, and the information shown on these pictures is to verify the effectiveness of the model? Thanks again for your reply.

**Comment by MingjunZhong** [comment]
@ZhuzengChu  MAE is more sensible for evaluating performance and thus more important.
Looking at both graph, the prediction is so bad - the model obviously was not well trained. I am afraid fridge is hard to train, but kettle is the easiest application to start with if you just started this project.
Anyway the results you show here does not look right.

**Comment by ZhuzengChu** [comment]
@MingjunZhong There is no problem with the MSE and SAE results obtained in the test set.  It is not clear why the effect is so poor.  It should be that I did not understand the code and some unknown errors occurred.  Finally, thanks for your guidance.

**Comment by MingjunZhong** [comment]
@ZhuzengChu  Show me these graphs after you have made it correct and let me know questions if you will have

**Comment by RengarWang** [comment]
yes , I had the same problem, the predicted value is always a straight line, i use light.
![light](https://user-images.githubusercontent.com/51586022/96246335-12a52780-0fdb-11eb-87f4-c8c58ee70987.png)
the truth is :
![light_truth](https://user-images.githubusercontent.com/51586022/96246424-31a3b980-0fdb-11eb-87b8-47f443c67716.png)
the predicted is :
![light_predicted](https://user-images.githubusercontent.com/51586022/96246485-4718e380-0fdb-11eb-9f83-26815422de8b.png)
the result is :
![1602843298(1)](https://user-images.githubusercontent.com/51586022/96246651-88a98e80-0fdb-11eb-96f6-19e2ac4ad319.png)

**Comment by MingjunZhong** [comment]
@RengarWang That looks the model was not well trained. Did you normalize the training data? How is the prediction when you apply the model to the **_training data_**? How was the performance when you train other appliances like the kettle?

**Comment by RengarWang** [comment]
> @RengarWang That looks the model was not well trained. Did you normalize the training data? How is the prediction when you apply the model to the **_training data_**? How was the performance when you train other appliances like the kettle?

Thank you. I've normalized it, but the result is still straight. I'm going to try something else now.

**Comment by LH1217** [comment]
> @RengarWang That looks the model was not well trained. Did you normalize the training data? How is the prediction when you apply the model to the **_training data_**? How was the performance when you train other appliances like the kettle?

Hello, I am a novice. I used the refit data set and encountered the same problem. How do I normalize the training data?

**Comment by RobinHan24** [comment]
@MingjunZhong @ZhuzengChu @RengarWang Hello, everyone. I just met the same problem. I did it according to the code. But my result is always a straight line. Could you please help me. Thanks a lot.

**Comment by RobinHan24** [comment]
@RengarWang @MingjunZhong @ZhuzengChu  the acc is always 0.00000
2021-03-01 13:07:44,444 [INFO ]  ------------------------- Epoch 6 of 30 --------------------------
2021-03-01 13:07:57,437 [INFO ]  loss_ep: 0.013152
2021-03-01 13:07:57,437 [INFO ]  Epoch 6 of 30 took 12.992529s
2021-03-01 13:07:57,437 [INFO ]  Validation...
2021-03-01 13:08:01,337 [INFO ]     train loss/n_batch_train: 0.012959
2021-03-01 13:08:01,337 [INFO ]     train loss: 0.129593, n_batch_train: 10
2021-03-01 13:08:05,357 [INFO ]      val loss: 0.033231
2021-03-01 13:08:05,357 [INFO ]      Current valid loss was 0.033231, acc was 0.000000, train loss was 0.012959, acc was 0.000000 at epoch 6.
2021-03-01 13:08:05,357 [INFO ]  ------------------------- Epoch 7 of 30 --------------------------
2021-03-01 13:08:18,588 [INFO ]  loss_ep: 0.012880
2021-03-01 13:08:18,588 [INFO ]  Epoch 7 of 30 took 13.230522s
2021-03-01 13:08:18,588 [INFO ]  Validation...
2021-03-01 13:08:22,523 [INFO ]     train loss/n_batch_train: 0.013669
2021-03-01 13:08:22,523 [INFO ]     train loss: 0.136689, n_batch_train: 10
2021-03-01 13:08:26,405 [INFO ]      val loss: 0.033339
2021-03-01 13:08:26,405 [INFO ]      Current valid loss was 0.033339, acc was 0.000000, train loss was 0.013669, acc was 0.000000 at epoch 7.

**Comment by GrandeVenti** [comment]
@RobinHan24 @ZhuzengChu @RengarWang hello, have you solved the problem that the prediction result is a straight line? I also encounter this problem and tried many methods. Thanks!

**Comment by suoyu2333** [comment]
> I also encounter this problem and tried many methods. Thanks!
I also encounter this problem .Have you solved the problem？I really want to know how to solve it.

**Comment by henriquebol** [comment]
> @RobinHan24 @ZhuzengChu @RengarWang hello, have you solved the problem that the prediction result is a straight line? I also encounter this problem and tried many methods. Thanks!

Same problem here

**Comment by DuoCaiXie** [comment]
> > @RobinHan24 @ZhuzengChu @RengarWang hello, have you solved the problem that the prediction result is a straight line? I also encounter this problem and tried many methods. Thanks!
> 
> Same problem here

hello,I also met this problem ,would you solve it?

**Comment by MingjunZhong** [comment]
What appliance are using working on? Can you use your trained model on your training data to see if it still output straight lines? Have you preprocess the test data using the same parameters on training data?

**Comment by DuoCaiXie** [comment]
> 
Hi，Thank you for your reply，I use the Kettle in ukdale and the washing machine in redd,However, the truth value of the image drawn after model training is always 0，So,I am very confused,And I did data preprocessing in accordance with your method.The image below is drawn by the washing machine：
<img width="541" alt="image" src="https://user-images.githubusercontent.com/61140297/161245871-d7e9fdae-7ac5-4ff2-a3f1-97febb3e3cfb.png">

**Comment by MingjunZhong** [comment]
Have you mapped back to predicted values using those parameters in preprocessing? What does it look like for kettle?

**Comment by DuoCaiXie** [comment]
> Have you mapped back to predicted values using those parameters in preprocessing? What does it look like for kettle?

<img width="463" alt="image" src="https://user-images.githubusercontent.com/61140297/161379194-af39d4b0-9417-4c2d-aea3-d379e2f0f5bb.png">

**Comment by DuoCaiXie** [comment]
> Have you mapped back to predicted values using those parameters in preprocessing? What does it look like for kettle?

I'm sorry for the late reply. I didn't understand what you meant. The above figure is the prediction curve of the kettle.

**Comment by ic085908** [comment]
I have the same problem with a straight line. 
I ran the default code without changing any parameters. Could you check it? @MingjunZhong 
![Figure 2022-06-19 184220](https://user-images.githubusercontent.com/46443600/174495709-9be01f1d-0803-4b52-b37c-2a482291eb05.png)

**Comment by MingjunZhong** [comment]
> 

Based on the results of kettle you have plotted, it was working very well indeed. That means your code should be fine.

**Comment by MingjunZhong** [comment]
> I have the same problem with a straight line. I ran the default code without changing any parameters. Could you check it? @MingjunZhong ![Figure 2022-06-19 184220](https://user-images.githubusercontent.com/46443600/174495709-9be01f1d-0803-4b52-b37c-2a482291eb05.png)

Do you have a better sequence in which the kettle was on? Can you firstly test the trained model on your training data to see the performance?

**Comment by MingjunZhong** [comment]
> > Have you mapped back to predicted values using those parameters in preprocessing? What does it look like for kettle?
> 
> <img alt="image" width="463" src="https://user-images.githubusercontent.com/61140297/161379194-af39d4b0-9417-4c2d-aea3-d379e2f0f5bb.png">

Based on the results of kettle you have plotted, it was working very well indeed. That means your code should be fine.

**Comment by ic085908** [comment]
> > > Have you mapped back to predicted values using those parameters in preprocessing? What does it look like for kettle?
> > 
> > 
> > <img alt="image" width="463" src="https://user-images.githubusercontent.com/61140297/161379194-af39d4b0-9417-4c2d-aea3-d379e2f0f5bb.png">
> 
> Based on the results of kettle you have plotted, it was working very well indeed. That means your code should be fine.

Thank you, I increased the number of rows of the dataset to take training data from that set default of 10000. So that can help to solve a straight line.
![image](https://user-images.githubusercontent.com/46443600/179359699-f84c1f52-7bc2-4548-9d33-b7823766bed4.png)

**Comment by wangskkk** [comment]
Turning the crop parameter in train_main.py to 2000000 while replacing test_directory in test_main.py with the training set improves the situation.

**Comment by MingjunZhong** [comment]
> Turning the crop parameter in train_main.py to 2000000 while replacing test_directory in test_main.py with the training set improves the situation.

That's great and thanks!
